ansible_ai_connect/ai/api/formatter.py

```diff
@@ -363,15 +363,36 @@ def get_task_names_from_tasks(tasks):
     return names
 
 
-def restore_original_task_names(output_yaml, prompt):
+def restore_original_task_names(output_yaml, prompt, payload_context=""):
     if output_yaml and is_multi_task_prompt(prompt):
+        full_yaml = payload_context + output_yaml
+        try:
+            data = yaml.safe_load(full_yaml)
+        except Exception as exc:
+            logger.exception(
+                f"Error while loading the result role/playbook YAML:{exc} "
+                f"for restoring the original task names"
+            )
+            return output_yaml
         prompt_tasks = get_task_names_from_prompt(prompt)
-        matches = re.finditer(r"^- name:\s+(.*)", output_yaml, re.M)
-        for i, match in enumerate(matches):
+        task_list = []
+        if isinstance(data, list) and len(data) > 0 and isinstance(data[0], dict):
+            # Tries to get the tasks from the dict if it's a playbook (payload_context!='')
```

Comment:
We can't assume that empty payload_context implies it's a tasks file in a role and non-empty payload_context implies playbook. The multi-task prompt, in either type of file, may not be the first task in the file. And when it's not, the logic here currently fails because it assumes the first task in task_list is the first task of the multi-prompt. Please add tests (and fix the logic here) for playbooks and tasks in a role where the multi-task prompt is after an existing task.